presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergAbstractMetadata.java

```diff
@@ -916,39 +903,29 @@ public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandl
             throw new TableNotFoundException(handle.getSchemaTableName());
         }
 
-        TableScan scan = icebergTable.newScan();
         TupleDomain<IcebergColumnHandle> domainPredicate = layoutHandle.getValidPredicate();
-
-        if (!domainPredicate.isAll()) {
-            Expression filterExpression = toIcebergExpression(domainPredicate);
-            scan = scan.filter(filterExpression);
-        }
-
-        try (CloseableIterable<FileScanTask> files = scan.planFiles()) {
-            return OptionalLong.of(removeScanFiles(icebergTable, files));
-        }
-        catch (IOException e) {
-            throw new PrestoException(GENERIC_INTERNAL_ERROR, "failed to scan files for delete", e);
-        }
+        return removeScanFiles(icebergTable, domainPredicate);
     }
 
     /**
-     * Deletes all the files within a particular scan
+     * Deletes all the files for a specific predicate
      *
      * @return the number of rows deleted from all files
      */
-    private long removeScanFiles(Table icebergTable, Iterable<FileScanTask> scan)
+    private OptionalLong removeScanFiles(Table icebergTable, TupleDomain<IcebergColumnHandle> predicate)
     {
         transaction = icebergTable.newTransaction();
-        DeleteFiles deletes = transaction.newDelete();
-        AtomicLong rowsDeleted = new AtomicLong(0L);
-        scan.forEach(t -> {
-            deletes.deleteFile(t.file());
-            rowsDeleted.addAndGet(t.estimatedRowsCount());
-        });
-        deletes.commit();
+        DeleteFiles deleteFiles = transaction.newDelete()
+                .deleteFromRowFilter(toIcebergExpression(predicate));
+        deleteFiles.commit();
         transaction.commitTransaction();
-        return rowsDeleted.get();
+
+        Map<String, String> summary = icebergTable.currentSnapshot().summary();
+        long deletedRecords = Long.parseLong(summary.getOrDefault(DELETED_RECORDS_PROP, "0"));
+        long removedPositionDeletes = Long.parseLong(summary.getOrDefault(REMOVED_POS_DELETES_PROP, "0"));
+        long removedEqualityDeletes = Long.parseLong(summary.getOrDefault(REMOVED_EQ_DELETES_PROP, "0"));
+        // Removed rows count is inaccurate when existing equality delete files
+        return OptionalLong.of(deletedRecords - removedPositionDeletes - removedEqualityDeletes);
```

Comment:
Is the assumption that records from position and equality deletes are counted twice? From my understanding I think it should be: `deletedRecords + removedPositionDeletes + removedEqualityDeletes`.

I was looking at the code here, and it doesn't seem like it would be double-counted based on this conditional since each deleted file would only add to its own counter.
https://github.com/apache/iceberg/blob/8d6bee736884575da7368e0963268d1cbe362d90/core/src/main/java/org/apache/iceberg/SnapshotSummary.java#L300-L321

I could be missing something though

Reply:
Thanks for the pointer to the code. As I understand, delete files are treated as normal data files when recording the summary, but they have different meanings when being applied to the rows in Iceberg table.

So the explanation about `deletedRecords - removedPositionDeletes - removedEqualityDeletes` is that, we removed data files that contains `deletedRecords` rows, of which `removedPositionDeletes + removedEqualityDeletes` rows have been already deleted through delete files.

For example, we have a data file contains 10 rows and a position delete files contains 2 rows. When we remove them all, we would get `deletedRecords = 10` and `removedPositionDeletes = 2` in current snapshot's summary. So we should get the deleted row count as 8.

Reply:
Thanks for the explanation. I also ran some more tests locally to confirm my understanding and your explanation matches 👍